### ISSUE TITLE
Color Emoji Font found but not displayed

### ISSUE THREAD

**Issue by Gordin** [polybar/polybar#1799]
## Describe the issue
Color Emoji are not shown in the bar, even when font is loaded correctly

**Expected behavior:**

"😱😠😒😊😃😈04. June 19, Tue" displayed in bar (Emojis in color)

**Actual behavior:**

With Noto Color Emoji installed:
"04. June 19, Tue" displayed in bar (No Emojis at all)

Without Noto Color Emoji installed:
"😱😠😒😊😃😈04. June 19, Tue" displayed in bar (Emojis in b/w, probably from Bitstream Vera Sans Mono as fallback)

## To Reproduce
*A minimal but complete config with which the problem occurs:*
```dosini
[bar/main]
monitor = eDP-1
modules-center = date
font-0 = "Mono:size=16:weight=regular;2"
font-1 = "Noto Color Emoji:pixelsize=16:style=Regular;2"

[module/date]
type = internal/date
interval = 1

date = "😱😠😒😊😃😈%d. %B %y, %a"
```
Noto Color Emoji should be installed

## Polybar Log
```
* Loading config: /home/aguth/.config/polybar/config_emojitest
- config: Current bar section: [bar/main]
* Loaded monitor eDP-1 (2560x1440+0+0)
* Bar geometry: 2560x24+0+0; Borders: 0,0,0,0
- bar: Attach X event sink
- bar: Attach signal receiver
- controller: Install signal handler
- controller: Setup user-defined modules
* Starting application
- controller: Main thread id = 1
* Starting module/date
- module/date: Thread id = 2
* Entering event loop (thread-id=1)
* Eventqueue worker (thread-id=3)
- bar: Create renderer
- renderer: Get TrueColor visual
- renderer: Allocate colormap
- renderer: Allocate output window
- renderer: Allocate window pixmaps
- renderer: Allocate graphic contexts
- renderer: Allocate alignment blocks
- renderer: Allocate cairo components
- renderer: Load fonts
* Configured DPI = 96x96
* Loaded font "Mono:size=16:weight=regular" (name=Bitstream Vera Sans Mono, offset=2, file=/usr/share/fonts/TTF/VeraMono.ttf)
* Loaded font "Noto Color Emoji:pixelsize=16:style=Regular" (name=Noto Color Emoji, offset=2, file=/usr/share/fonts/noto/NotoColorEmoji.ttf)
* Bar window: 0x1c00001
- bar: Reconfigure window
- bar: Set window WM_NAME
- bar: Set window _NET_WM_WINDOW_TYPE
- bar: Set window _NET_WM_STATE
- bar: Set window _NET_WM_DESKTOP
- bar: Set window _NET_WM_PID
- bar: Map window
- bar: Draw empty bar
- bar: Setup tray manager
* Disabling tray manager (reason: missing `tray-position`)
* module/date: Rebuilding cache
- bar: Force update
* Redrawing bar window
- renderer: flush(2 geom=208x24+1176+0, falloff=0)
- bar: Received expose event
```

## Screenshots
Bar with Noto Color Emoji installed and loaded:
![Bar with Noto Color Emoji](https://user-images.githubusercontent.com/118403/58904395-b607dc00-8707-11e9-9110-283506619704.png)
Bar without Noto Color Emoji installed:
![Bar without Noto Color Emoji](https://user-images.githubusercontent.com/118403/58904418-c324cb00-8707-11e9-89d2-300d9d835368.png)

## Environment:
* WM: i3
* Distro: Arch
* Output of `polybar -vvv`:
```
polybar 3.3.0-97-g5be532c5

Features: +alsa +curl +i3 +mpd +network(libnl) +pulseaudio +xkeyboard

X extensions: +randr (+monitors) +composite +xkb +xrm +xcursor

Build type: Release
Compiler: /usr/lib/ccache/bin/c++
Compiler flags: -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fno-plt -Wall -Wextra -Wpedantic -O3 -DNDEBUG
Linker flags: -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now
```
(This is polybar-git, but the current version from the repository has the same problem)
## Additional context

Showing Color Emojis works everywhere else in my System (Firefox, terminals, GTK) without problems. I also tried to use EmojiOne instead of Noto, but everything is the same. (Doesn't show in polybar, works everywhere else)

**Comment by Lomadriel** [comment]
```ini
[bar/main]
monitor = eDP-1
modules-center = date
font-0 = "Mono:size=16:weight=regular;2"
font-1 = "Noto Color Emoji:pixelsize=16:style=Regular;2"

[module/date]
type = internal/date
interval = 1

date = "😱😠😒😊😃😈%d. %B %y, %a"
```

In your config the module date use the font-0 Mono which have the Emoji but in b&w.
So if you want to display the Emoji with the "Noto Color Emoji" font you must use something like that:

```
[module/date]
type = internal/date
interval = 5

format-prefix = "😱😠😒😊😃😈"
format-prefix-font = 2

time = %H:%M
time-alt = %H:%M:%S
```

I used the "format-prefix" to avoid using the font for displaying the time because the number in Noto Color Emoji are really ugly :(

I also recommend you to add the scale property to your font to avoid a known issue (https://github.com/polybar/polybar/wiki/Known-Issues#huge-emojis):
```
font-1 = "Noto Color Emoji:pixelsize=16:style=Regular:scale=10;2"
```

Closing since this is not an issue.

**Comment by patrick96** [comment]
It's weird that no emojis are displayed at all if Noto Color Emoji is installed. I was not able to reproduce this but I'll take a closer look later.

**Comment by Lomadriel** [comment]
> It's weird that no emojis are displayed at all if Noto Color Emoji is installed. I was not able to reproduce this but I'll take a closer look later.

Oops, did not see that sorry :/

**Comment by Gordin** [comment]
I tried this config now to make sure polybar knows what font to choose, but I think polybar knew to choose the right font before, it's just not displaying any characters from it:
```dosini
[bar/main]
monitor = eDP-1
modules-center = date
font-0 = "Mono:size=16:weight=regular;2"
font-1 = "Noto Color Emoji:pixelsize=16:style=Regular;2"

[module/date]
type = internal/date
interval = 5

format-prefix = "😱😠😒😊😃😈"
format-prefix-font = 2

date = "%{T2}😱😠😒😊😃😈%{T-}%d. %B %y, %a"
```

Same result as before, no Emojis at all with Color Emojis installed

**Comment by Lomadriel** [comment]
Hum, that's really strange :/
Have you ever had this problem in a previous version of polybar?
I am also unable to reproduce this :/

**Comment by patrick96** [comment]
@Gordin could you try

```dosini
font-1 = "Noto Color Emoji:style=Regular:scale=10;2"
```

(I removed the `pixelsize` property)

**Comment by Gordin** [comment]
@Lomadriel I had it working a couple of days ago actually. I upgraded from 3.3.0 to 3.3.1 recently, the last upgrade before was last year. I downgraded to 3.3.0 to test, but same result.

I saw this during the build process though (during generation of the build files):
```
-- xpp: including xcb proto composite.xml
-- xpp: including xcb proto randr.xml
-- xpp: including xcb proto sync.xml
-- xpp: including xcb proto xkb.xml
-- xpp: including xcb proto xproto.xml
-- Font not found: fixed:pixelsize=10
-- Found font: unifont:fontformat=truetype
-- Font not found: siji:pixelsize=10
-- Configuring done
-- Generating done
```
Could the font not found be anything to go by? I have siji fonts installed. It's not in fc-list though, but I'm not sure if it even should be, as its a bdf font? Or is my fontconfig broken and it should be found?

@patrick96 Tried these, still no Emoji shown, font always loaded correctly in logfile
```
font-1 = "Noto Color Emoji:style=Regular:scale=10;2"
font-1 = "Noto Color Emoji"
font-1 = "Noto Color Emoji:size=20style=Regular:scale=10;2"
font-1 = "Noto Color Emoji:style=Regular;2"
```

**Comment by Lomadriel** [comment]
> Could the font not found be anything to go by?

This doesn't change anything if I am not wrong.

> Or is my fontconfig broken and it should be found?

Maybe you should try to run `sudo fc-cache -f -v` to rebuild you cache, however since it works on your other application, it shouldn't fix the problem but it's worth a try ^^

**Comment by patrick96** [comment]
> This doesn't change anything if I am not wrong.

Correct. This is only relevant for what font patterns are added to the sample config.

As I see it, since there is actually space reserved for the glyphs, there is something wrong during the actual rendering. So polybar finds the glyphs just find just when rendering, they're either not rendered or maybe rendered somewhere offscreen.

@Gordin can you try to use higher values for `scale` and see if the gap here gets smaller:

![](https://user-images.githubusercontent.com/118403/58904395-b607dc00-8707-11e9-9110-283506619704.png)

This would confirm that polybar actually finds the glyphs in "Noto Color Emoji".

Could you also play around with different offsets (the number after the `;`), to test whether or not the glyphs are just rendered somewhere offscreen.

**Comment by Gordin** [comment]
@patrick96 I tried a couple values, and also played around with colors, when I set fg and bg to red, if I only wrap that around the emoji, net a single red pixel is there.

I figured out the problem though, and I guess it's not really a bug, depending on what the expected behavior of fontconfig is. I saw that I also had problems in Firefox with Emoji *sometimes*. Most emoji would display fine, but some would render but would have 0 width with respect to other characters, so this `🌙⚡X` would display as just X, but with both Emoji stacked behind each other visible behind the X. I found out I actually had another fontconfig config file I didn't know about in my home folder (I usually just use /etc/...) and I think I set that one up some time ago with the help of this post: [Colorful emojis on Linux](https://github.com/wireapp/wire-desktop/wiki/Colorful-emojis-on-Linux)

I had the `<rejectfont>` part set to an Emoji Font **that I didn't have installed any more** and it never caused problems before because I must have had multiple emoji fonts and maybe just rejected one of them and my programs just used the next best one, but in Polybar only the fonts that I specifically loaded are available. So because I was trying to get a cleaner behavior for Emoji fonts and uninstalled all Emoji fonts but one, I guess fontconfig would always reject the one font I had installed because it had "Emoji" somewhere in the name...

So... I definitely fucked up with my config, but I feel like there there has to be a bug here, but probably in fontconfig and/or Firefox, stacking Glyphs of rejected Fonts behind other stuff does not seem like intended behavior. For Polybar, not displaying anything at all when a font is rejected actually seems like the right thing to do, but maybe this case should also trigger a `Dropping unmatched character`, or some kind of `Failed to load rejected font` message.

I'm closing this now, I'm not sure if you would consider the missing debug messages a bug, but they don't have anything to do with Emoji. Sorry for wasting your time with my broken config files...
